### PR TITLE
UI: pki configuration edit form

### PR DESCRIPTION
**Pull request by hellobontempo**
Adds pki edit configuration form
edit-config

### CODE REVIEW

File: ui/lib/pki/addon/components/page/pki-configuration-edit.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  @action
+  cancel() {
+    this.router.transitionTo('vault.cluster.secrets.backend.pki.configuration.index');
+  }
+
+  @action
+  handleTtl(attr: FormField, e: TtlEvent) {
+    const { enabled, goSafeTimeString } = e;
+    const ttlAttr = attr.name;
+    this.args.crl[ttlAttr as keyof PkiCrlTtls] = goSafeTimeString;
```
Comment (kiannaquach) [inline ui/lib/pki/addon/components/page/pki-configuration-edit.ts:67]: I learned something new here! I didn't have the chance to `as keyof` yet in Typescript.

File: ui/lib/pki/addon/routes/configuration/edit.js [hunk 2]
```diff
@@ -4,5 +4,21 @@
  */
 
 import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+// import { withConfirmLeave } from 'core/decorators/confirm-leave';
 
-export default class PkiConfigurationEditRoute extends Route {}
+// @withConfirmLeave()
```
Comment (kiannaquach) [inline ui/lib/pki/addon/routes/configuration/edit.js:10]: `withConfirmLeave` is commented out here. Is that intentional?

File: ui/lib/pki/addon/routes/configuration.js [hunk 3]
```diff
@@ -4,5 +4,48 @@
  */
 
 import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+import { hash } from 'rsvp';
+import { withConfig } from 'pki/decorators/check-config';
 
-export default class PkiConfigurationRoute extends Route {}
+@withConfig()
```
Comment (kiannaquach) [inline ui/lib/pki/addon/routes/configuration.js:11]: nice job moving this out from `configuration/index.js` to the parent level.

**Review by kiannaquach — DISMISSED** [review]
Awesome job, Claire!! 🎊

I had one question about permissions:
- If the user doesn't have permission to view or edit CRL's or URL's, should we hide the **Edit Configuration** button or hide the parts they have access rights to edit?

File: ui/app/models/pki/crl.js [hunk 4]
```diff
@@ -4,15 +4,80 @@
...
+    helperTextEnabled: 'Vault will rebuild the CRL in the below grace period before expiration',
+    helperTextDisabled: 'Vault will not automatically rebuild the CRL',
+  })
+  autoRebuildGracePeriod;
+
+  @attr('boolean') enableDelta; // add validations auto_rebuild must be enabled
+  @attr('string', {
+    label: 'Delta CRL building on',
+    labelDisabled: 'Delta CRL building off',
+    defaultValue: '15mh',
```
Comment (kiannaquach) [inline ui/app/models/pki/crl.js:36]: Is this supposed to be `15h` instead of `15mh` ?

File: ui/lib/pki/addon/components/page/pki-configuration-edit.hbs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        </button>
+        <button
+          {{on "click" this.cancel}}
+          type="button"
+          class="button has-left-margin-s"
+          data-test-configuration-edit-cancel
+        >
+          Cancel
+        </button>
+      </div>
```
Comment (kiannaquach) [inline ui/lib/pki/addon/components/page/pki-configuration-edit.hbs:93]: This probably isn't necessary since you're transitioning to the configuration page on submit, but do you want to show disabled/loading states?

File: ui/lib/pki/addon/routes/configuration/edit.js (hunk 2 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/routes/configuration/edit.js:10]: Nope! Thank you for the reminder, when I finished developing I want to comment it back in!

File: ui/lib/pki/addon/routes/configuration.js (hunk 3 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/routes/configuration.js:11]: It made sense where you had it before! But now that there's a child route sharing some resources, I moved it up. I wondered about splitting things up a bit here so only the shared data is in this route (crls and url fetches) which would mean moving `hasConfig` back down to the index. 

The pro would be we'd only be passing necessary data, but required model hooks in both the `pki/configuration/index.js` and `pki/configuration/edit.js` to feel a bit piecemeal. But this seemed like it might make it harder to track data down later, because we'd have to reference both the parent's model and child's - what are your thoughts?

File: ui/app/models/pki/crl.js (hunk 4 above)
Comment (hellobontempo) [reply ui/app/models/pki/crl.js:36]: it should be 15m actually! good catch!

**Comment by hellobontempo** [comment]
> Awesome job, Claire!! 🎊
> 
> I had one question about permissions:
> 
> * If the user doesn't have permission to view or edit CRL's or URL's, should we hide the **Edit Configuration** button or hide the parts they have access rights to edit?

Good question! I followed the pattern I saw in the [pki generate root template](https://github.com/hashicorp/vault/blob/9de1cfcf8279c07135b4d8ad06511b0590cb47e1/ui/lib/pki/addon/components/pki-generate-root.hbs#L76) that shows an empty state if the user doesn't have permission to set urls.

And wrapped both `fieldset` elements in `canSet` conditionals [here for urls](https://github.com/hashicorp/vault/blob/54a988b6c11f1f7a29b8135f88e8a6735d24dcb9/ui/lib/pki/addon/components/page/pki-configuration-edit.hbs#L10) and [here for crl](https://github.com/hashicorp/vault/blob/54a988b6c11f1f7a29b8135f88e8a6735d24dcb9/ui/lib/pki/addon/components/page/pki-configuration-edit.hbs#L26) 

I opted for this instead of hiding the button, but open to doing otherwise!
Screenshot 2023-04-20 at 9 46 20 AM

File: ui/lib/pki/addon/routes/configuration.js (hunk 3 above)
Comment (kiannaquach) [reply ui/lib/pki/addon/routes/configuration.js:11]: I just realized this - by adding this data (`crls and urls`) in the configuration parent then the tidy and create pages will also try to fetch that data? If that's the case, I'm wondering if it should be shared with all the configuration pages since create + tidy doesn't need all of this data?

Comment (hellobontempo) [reply ui/lib/pki/addon/routes/configuration.js:11]: They won't re-fetch the data, but yes the data is passed down. Which is why I was considering intercepting in the model hooks and only passing the necessary data. However, then it got a little hard to follow which data was coming from where. For example, you'd have to look at the configuration parent route beside the child route to look at the combined model hook data. So I opted to keep it all in the parent, since it's not a ton of extra data. But open to considering alternatives!

File: ui/lib/pki/addon/components/page/pki-configuration-edit.hbs (hunk 5 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/components/page/pki-configuration-edit.hbs]: good idea!

File: ui/lib/pki/addon/routes/configuration.js (hunk 3 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/routes/configuration.js]: @kiannaquach ...after all that back and fourth I ended up adding the model hooks back to the child routes 😅. A test was failing because of the issuerModel created just for capabilities and when I revisited this, I realized it actually was easier to track down the data when each route had it's model hook clearly defined. Since I changed my mind, I wanted to follow up and let you know!

File: ui/lib/pki/addon/routes/configuration/create.js [hunk 6]
```diff
@@ -28,12 +28,4 @@ export default class PkiConfigurationCreateRoute extends Route {
       { label: 'configure' },
     ];
   }
-
-  async getOrCreateUrls(backend) {
```
Comment (hellobontempo) [inline ui/lib/pki/addon/routes/configuration/create.js:32]: I think only ever need to findRecord because as soon as the pki mount is mounted, we can `READ` the `config/urls` endpoint. The response below is from a freshly mounted pki engine called `pki-unconfig`.
```sh
~|⇒ curl \
    --header "X-Vault-Token: root" \
    http://127.0.0.1:8200/v1/pki-unconfig/config/urls | jq

# --  response --
{
  "request_id": "e440be87-6a56-9c1d-7a62-9b78876792d9",
  "lease_id": "",
  "renewable": false,
  "lease_duration": 0,
  "data": {
    "crl_distribution_points": [],
    "enable_templating": false,
    "issuing_certificates": [],
    "ocsp_servers": []
  },
  "wrap_info": null,
  "warnings": null,
  "auth": null
}

File: ui/app/models/pki/crl.js [hunk 7]
```diff
@@ -4,15 +4,84 @@
...
+  @attr('string') deltaRebuildInterval;
+
+  @attr('object', {
+    label: 'Auto-rebuild on',
+    labelDisabled: 'Auto-rebuild off',
+    editType: 'ttl',
+    defaultValue() {
+      return { enabled: false, duration: '12h' };
+    },
+    helperTextEnabled: 'Vault will rebuild the CRL in the below grace period before expiration',
```
Comment (Monkeychip) [inline ui/app/models/pki/crl.js:31]: Probably a design question, but I'm seeing some helperText with periods and some without. I know in general we're a bit all over with this, but thought I would point it out.

File: ui/app/models/pki/crl.js [hunk 8]
```diff
@@ -4,15 +4,84 @@
...
+    labelDisabled: 'OCSP responder APIs disabled',
+    defaultValue() {
+      return { enabled: true, duration: '12h' };
+    },
+    helperTextEnabled: "Requests about a certificate's status will be valid for:",
+    helperTextDisabled: 'Requests cannot be made to check if an individual certificate is valid.',
+  })
+  ocspExpiryData; // sets ocsp_disable (boolean), ocsp_expiry (duration)
+
+  // TODO missing from designs, enterprise only - add?
```
Comment (Monkeychip) [inline ui/app/models/pki/crl.js:71]: Is there a corresponding ticket for this todo? Making sure it doesn't get lost.

File: ui/app/serializers/pki/crl.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import ApplicationSerializer from '../application';
+
+// The object keys ('enabled', 'duration') in the pki/crl model map to an API parameter
+// Each object is updated by a single ttl input in the pki-configuration-edit form
+// In some cases 'false' means enabled (i.e. { disable: false } => ttl field is toggled ON)
```
Comment (Monkeychip) [inline ui/app/serializers/pki/crl.js:10]: great comment!

File: ui/lib/pki/addon/routes/configuration/edit.js [hunk 10]
```diff
@@ -4,5 +4,30 @@
  */
 
 import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+import { hash } from 'rsvp';
+import { withConfirmLeave } from 'core/decorators/confirm-leave';
```
Comment (Monkeychip) [inline ui/lib/pki/addon/routes/configuration/edit.js:9]: great use of a decorator!

File: ui/lib/pki/addon/templates/configuration/edit.hbs [hunk 11]
```diff
@@ -1 +1,13 @@
-configuration.edit
```
Comment (Monkeychip) [inline ui/lib/pki/addon/templates/configuration/edit.hbs:1]: very cool to see this part of the project getting tackled. It's a big step towards completion.

**Review by Monkeychip — APPROVED** [review]
Great work! 🥳

File: ui/app/adapters/pki/urls.js [hunk 12]
```diff
@@ -13,13 +13,12 @@ export default class PkiUrlsAdapter extends ApplicationAdapter {
...
-    return this._url(snapshot.record.id);
+  updateRecord(store, type, snapshot) {
+    const data = snapshot.serialize();
+    return this.ajax(this._url(snapshot.record.id), 'POST', { data });
   }
+
   urlForFindRecord(id) {
     return this._url(id);
   }
-  urlForUpdateRecord(store, type, snapshot) {
```
Comment (hellobontempo) [inline ui/app/adapters/pki/urls.js:22]: this was using `PUT`

File: ui/app/models/pki/crl.js (hunk 8 above)
Comment (hellobontempo) [reply ui/app/models/pki/crl.js:71]: Good call, I was planning to do in this PR but never got confirmation. I'll make a ticket

File: ui/app/models/pki/crl.js (hunk 7 above)
Comment (hellobontempo) [reply ui/app/models/pki/crl.js:31]: Just double checked and these are all copy and pasted from the designs - probably worth following up with the latest iteration of the design team and see how we want to handle!

File: ui/lib/pki/addon/decorators/check-config.js [hunk 13]
```diff
@@ -21,6 +22,7 @@ export function withConfig() {
       return SuperClass;
     }
     return class CheckConfig extends SuperClass {
+      @service secretMountPath;
```
Comment (hellobontempo) [inline ui/lib/pki/addon/decorators/check-config.js:25]: this okay to do here?

File: ui/lib/pki/addon/routes/configuration.js [hunk 14]
```diff
@@ -4,5 +4,34 @@
...
+    }
+  }
+
+  async fetchCrl(backend) {
+    try {
+      return await this.store.findRecord('pki/crl', backend);
+    } catch (e) {
+      return e.httpStatus;
+    }
+  }
```
Comment (hellobontempo) [inline ui/lib/pki/addon/routes/configuration.js:27]: moved up from `configuration/index.js` route

File: ui/app/models/pki/crl.js [hunk 15]
```diff
@@ -28,11 +28,12 @@ export default class PkiCrlModel extends Model {
...
 
-  // edit form ttl attrs
-  @attr('object', {
+  @attr('ttl', {
+    enabledKey: 'auto_rebuild',
+    durationKey: 'auto_rebuild_grace_period',
+    isOppositeValue: false,
     label: 'Auto-rebuild on',
     labelDisabled: 'Auto-rebuild off',
-    editType: 'ttl',
```
Comment (zofskeez) [inline ui/app/models/pki/crl.js]: This would still be required to show the `TtlPicker` in the `FormField` component but it might be better to add a condition to also look for the ttl type. What do you think?
```hbs
{{else if (or (eq @attr.type "ttl") (eq @attr.options.editType "ttl"))}}
```

File: ui/app/serializers/pki/crl.js [hunk 16]
```diff
@@ -9,56 +9,27 @@ import ApplicationSerializer from '../application';
...
       if (key in json) {
-        const { enabledKey, durationKey, isOppositeValue } = MODEL_TO_API_PARAMS[key];
-        const { enabled, duration } = json[key];
-        json[enabledKey] = isOppositeValue ? !enabled : enabled;
-        json[durationKey] = duration;
+        json = { ...json, ...json[key] };
         delete json[key];
       }
-    }
+    });
```
Comment (zofskeez) [inline ui/app/serializers/pki/crl.js]: You should be able to remove this once the ttl transform is used on all of the necessary attributes since it will also serialize the data into the expected payload.

File: ui/app/models/pki/crl.js (hunk 15 above)
Comment (hellobontempo) [reply ui/app/models/pki/crl.js]: I'm not using the form field component for this model, and this model only appears in one form

File: ui/app/adapters/pki/urls.js [hunk 17]
```diff
@@ -13,13 +13,12 @@ export default class PkiUrlsAdapter extends ApplicationAdapter {
     return `${this.buildURL()}/${encodePath(backend)}/config/urls`;
   }
 
-  urlForCreateRecord(modelName, snapshot) {
```
Comment (hellobontempo) [inline ui/app/adapters/pki/urls.js:16]: we only ever `findRecord`

File: ui/app/models/pki/crl.js [hunk 18]
```diff
@@ -4,15 +4,67 @@
...
 export default class PkiCrlModel extends Model {
   // This model uses the backend value as the model ID
-  get useOpenAPI() {
-    return true;
-  }
 
-  @attr('string') expiry;
   @attr('boolean') autoRebuild;
-  @attr('string') ocspExpiry;
+  @attr('string', {
```
Comment (hellobontempo) [inline ui/app/models/pki/crl.js:15]: I considered making these `ttls` and creating a new transform, but I didn't see the immediate value in doing that. But might be a nice follow on, to have an attribute option that chose which key of the ttl object (i.e. `goSafeTimeString`) the att corresponds to

File: ui/lib/pki/addon/decorators/check-config.js (hunk 13 above)
Comment (zofskeez) [reply ui/lib/pki/addon/decorators/check-config.js:25]: Yes I think that is good, otherwise you rely on the parent class to inject the service.

File: ui/lib/pki/addon/routes/configuration.js [hunk 19]
```diff
@@ -4,5 +4,34 @@
...
+
+  model() {
+    const engine = this.modelFor('application');
+    return hash({
+      engine,
+      urls: this.fetchUrls(engine.id),
+      crl: this.fetchCrl(engine.id),
```
Comment (zofskeez) [inline ui/lib/pki/addon/routes/configuration.js:34]: This is all good but just a suggestion to simplify, you could return the promise with a catch. For example:
```js
urls: this.store.findRecord('pki/crl', engine.id).catch((e) => e.httpStatus)
```

File: ui/lib/pki/addon/routes/configuration/edit.js [hunk 20]
```diff
@@ -4,5 +4,30 @@
...
+export default class PkiConfigurationEditRoute extends Route {
+  @service secretMountPath;
+
+  model() {
+    const { urls, crl, engine } = this.modelFor('configuration');
+    return hash({
+      engineId: engine.id,
+      urls,
+      crl,
+    });
```
Comment (zofskeez) [inline ui/lib/pki/addon/routes/configuration/edit.js:21]: I'm not sure you would need to use `hash` here. Since the configuration route is the parent the model should already be loaded. Although since you used hash there I wonder if `urls` and `crl` are promises? This is more of a question than anything 😄

File: ui/lib/pki/addon/components/page/pki-configuration-details.hbs [hunk 21]
```diff
@@ -38,14 +38,37 @@
     <h2 class="title is-4 has-bottom-margin-xs has-top-margin-xl has-border-bottom-light has-bottom-padding-s">
       Certificate Revocation List (CRL)
     </h2>
-    <InfoTableRow @label="Expiry" @value={{@crl.expiry}} />
-    <InfoTableRow @label="Auto-rebuild" @value={{if @crl.autoRebuild "On" "Off"}} />
-
+    <InfoTableRow @label="CRL building" @value={{if @crl.disable "Disabled" "Enabled"}} />
```
Comment (kiannaquach) [inline ui/lib/pki/addon/components/page/pki-configuration-details.hbs:41]: thanks so much for updating this! love that it creates more consistency in the app.

**Review by zofskeez — APPROVED** [review]
I added a few suggestions but LGTM. Thanks!

File: ui/lib/pki/addon/components/page/pki-configuration-edit.hbs [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+            disabled={{this.save.isRunning}}
+            data-test-configuration-edit-save
+          >
+            Save
+          </button>
+        {{/if}}
+        <button
+          {{on "click" this.cancel}}
+          type="button"
+          class="button has-left-margin-s"
```
Comment (kiannaquach) [inline ui/lib/pki/addon/components/page/pki-configuration-edit.hbs:79]: when the form is submitting or `this.save.isRunning`, do we want to disable the button?

File: ui/lib/pki/addon/components/page/pki-configuration-edit.hbs [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+        Certificate Revocation List (CRL)
+      </h2>
+      {{#if @crl.canSet}}
+        {{#each @crl.formFields as |attr|}}
+          {{#if (eq attr.name "ocspExpiry")}}
+            <h2 class="title is-size-5 has-border-bottom-light page-header">
+              Online Certificate Status Protocol (OCSP)
+            </h2>
+          {{/if}}
+          {{#if (or (includes attr.name this.alwaysRender) (not @crl.disable))}}
```
Comment (kiannaquach) [inline ui/lib/pki/addon/components/page/pki-configuration-edit.hbs:36]: `includes` is new! I haven't had the change to use that!

File: ui/lib/pki/addon/decorators/check-config.js (hunk 13 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/decorators/check-config.js:25]: yep! 

and exactly @zofskeez - it seems like if this is where it's called it made sense to have

**Review by kiannaquach — APPROVED** [review]
Your PR lgtm! Nice job with PR and appreciate the refactor you did here 😊

File: ui/lib/pki/addon/routes/configuration/edit.js (hunk 20 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/routes/configuration/edit.js:21]: Nope, those should just be record from the parent model,

File: ui/lib/pki/addon/components/page/pki-configuration-edit.hbs (hunk 22 above)
Comment (hellobontempo) [reply ui/lib/pki/addon/components/page/pki-configuration-edit.hbs:79]: thank you!
